include/pagmo/problems/golomb_ruler.hpp

```diff
@@ -0,0 +1,216 @@
+ * \f[
+ *   \begin{array}{rl}
+ *   \mbox{find:} & 1 \le d_i \le l_{max}, \forall i=1..n-1 \\
+ *   \mbox{to minimize: } & \sum_i d_i  \\
+ *   \mbox{subject to:} & |a_i-a_j| \neq |a_l - a_m|, \forall (\mbox{distinct}) i,j,l,m \in [0, n]
+ * \end{array}
+ * \f]
+ *
+ * We transcribe the constraints as one single equality constraint: \f$ c = 0 \f$ where \f$ c\f$ is the count of
+ * repeated distances.
+ *
+ * See: https://en.wikipedia.org/wiki/Golomb_ruler
+ *
+ */
+class golomb_ruler
+{
+public:
+    /// Constructor from ruler order and maximal consecutive ticks distance
+    /**
+     * Constructs a Golomb Ruler problem.
+     *
+     * @param order ruler order.
+     * @param upper_bound maximum distance between consecutive ticks.
+     *
+     * @throw std::invalid_argument if \p order is < 2 or \p upper_bound is < 1.
+     * @throw std::overflow_error if \p upper_bound is too large.
+     */
+    golomb_ruler(unsigned order = 3u, unsigned upper_bound = 10) : m_order(order), m_upper_bound(upper_bound)
+    {
+        if (order < 2u) {
+            pagmo_throw(std::invalid_argument, "Golomb ruler problem must have at least order 2, while "
+                                                   + std::to_string(order) + " was requested.");
+        }
+        if (upper_bound < 2u) {
+            pagmo_throw(
+                std::invalid_argument,
+                "The upper bound for the maximum distance between consecutive ticks has to be at least 2, while "
+                    + std::to_string(upper_bound) + " was requested.");
+        }
+        // Overflow can occur when evaluating the fitness later if the upper_bound is too large.
+        if (static_cast<double>(upper_bound) * (order - 1u)
+            > static_cast<double>(std::numeric_limits<unsigned>::max())) {
+            pagmo_throw(
+                std::overflow_error,
+                "Overflow in Golomb ruler problem, select a smaller maximum distance between consecutive ticks.");
+        }
+    }
+    /// Fitness computation
+    /**
+     * Computes the fitness for this UDP
+     *
+     * @param x the decision vector.
+     *
+     * @return the fitness of \p x.
+     */
+    vector_double fitness(const vector_double &x) const
+    {
+        vector_double f(2, 0.);
+        // 1 - We compute the ticks (the ruler length will be the last tick since we start from 0)
+        vector_double ticks(x.size() + 1, 0.);
+        std::partial_sum(x.begin(), x.end(), ticks.begin() + 1u);
+        f[0] = ticks.back();
+        // 2 - We compute all pairwise distances
+        vector_double distances;
+        for (decltype(ticks.size()) i = 0; i < ticks.size() - 1; ++i) {
+            for (decltype(ticks.size()) j = i + 1; j < ticks.size(); ++j) {
+                auto tmp = ticks[j] - ticks[i];
+                distances.push_back(tmp);
+            }
+        }
+        // 3 - We compute how many duplicate distances are there.
+        std::sort(distances.begin(), distances.end());
```

Comment:
Here we have the usual issue of potentially sorting with ``NaN`` values. Perhaps we could add a check that the produced distances are all non-``NaN``? Or otherwise we can use one of the custom comparators we have in some utility file (in which case, such a custom comparator should also be used in the ``std::unique()`` call below).